### PR TITLE
Feature/message row

### PR DESCRIPTION
**Pull request by bmarty**
- Improve performance of message adapter by keeping what we need from RoomState in MessageRow
- Fixes vector-im/riot-android#2639

### CODE REVIEW

File: matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/MXEventTimeline.java [hunk 1]
```diff
@@ -470,10 +471,10 @@ private void addPaginationEvents(List<Event> events,
             for (Event stateEvent : stateEvents) {
                 if (direction == Direction.BACKWARDS) {
                     // Enrich the timeline root state with the additional state events observed during back pagination
-                    processStateEvent(stateEvent, Direction.FORWARDS);
+                    processStateEvent(stateEvent, Direction.FORWARDS, true);
                 }
 
-                processStateEvent(stateEvent, direction);
+                processStateEvent(stateEvent, Direction.BACKWARDS, true);
```
Comment (giomfo) [inline matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/MXEventTimeline.java]: There is an issue here when direction is Forward, we never call processStateEvent

File: matrix-sdk/src/main/java/org/matrix/androidsdk/adapters/MessageRow.java [hunk 2]
```diff
@@ -35,8 +54,12 @@
     public MessageRow(Event event, RoomState roomState) {
         mEvent = event;
         mRoomState = roomState;
-    }
 
+        mUserDisplayName = (null == roomState) ? event.getSender() : roomState.getMemberName(event.getSender());
+        mRoomCreateContent = (null == roomState) ? null : roomState.getRoomCreateContent();
+        mSender = roomState.getMember(event.getSender());
```
Comment (giomfo) [inline matrix-sdk/src/main/java/org/matrix/androidsdk/adapters/MessageRow.java]: 1. What happen if roomState is null?
2. A deep copy of the room member should be useful here. Then mUserDisplayName should be useless

File: matrix-sdk/src/main/java/org/matrix/androidsdk/data/RoomState.java [hunk 3]
```diff
@@ -946,7 +951,7 @@ public boolean applyState(IMXStore store, Event event, EventTimeline.Direction d
                         // duplicated message ?
                         if (member.equals(currentMember)) {
                             Log.e(LOG_TAG, "## applyState() : seems being a duplicated event for " + userId + " in room " + roomId);
-                            return false;
+                            // return false;
```
[the remark this answers is not in the document]
Comment (bmarty) [reply matrix-sdk/src/main/java/org/matrix/androidsdk/data/RoomState.java]: Done

File: matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/MXEventTimeline.java (hunk 1 above)
Comment (bmarty) [reply matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/MXEventTimeline.java]: Yes, thx

File: matrix-sdk/src/main/java/org/matrix/androidsdk/adapters/MessageRow.java (hunk 2 above)
Comment (bmarty) [reply matrix-sdk/src/main/java/org/matrix/androidsdk/adapters/MessageRow.java]: Dealing with point 1 makes point 2 non applicable

File: matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/MXEventTimeline.java [hunk 4]
```diff
@@ -482,8 +483,10 @@ private void addPaginationEvents(List<Event> events,
             boolean processedEvent = true;
 
             if (event.stateKey != null) {
+                boolean considerNewContent = direction == Direction.FORWARDS;
+
                 deepCopyState(direction);
-                processedEvent = processStateEvent(event, direction);
+                processedEvent = processStateEvent(event, direction, considerNewContent);
```
Comment (giomfo) [inline matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/MXEventTimeline.java]: Why don't you "enrich the timeline root state with the additional state events observed during back pagination" here?

File: matrix-sdk/src/main/java/org/matrix/androidsdk/data/RoomState.java [hunk 5]
```diff
@@ -872,17 +871,23 @@ public String encryptionAlgorithm() {
     /**
      * Apply the given event (relevant for state changes) to our state.
      *
-     * @param store     the store to use
-     * @param event     the event
-     * @param direction how the event should affect the state: Forwards for applying, backwards for un-applying (applying the previous state)
+     * @param store              the store to use
+     * @param event              the event
+     * @param considerNewContent how the event should affect the state: true for applying, false for un-applying (applying the previous state)
```
[the remark this answers is not in the document]
Comment (bmarty) [reply matrix-sdk/src/main/java/org/matrix/androidsdk/data/RoomState.java]: Comment updated

File: matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/MXEventTimeline.java [hunk 6]
```diff
@@ -470,10 +471,10 @@ private void addPaginationEvents(List<Event> events,
             for (Event stateEvent : stateEvents) {
                 if (direction == Direction.BACKWARDS) {
                     // Enrich the timeline root state with the additional state events observed during back pagination
-                    processStateEvent(stateEvent, Direction.FORWARDS);
+                    processStateEvent(stateEvent, Direction.FORWARDS, true);
                 }
 
-                processStateEvent(stateEvent, direction);
+                processStateEvent(stateEvent, direction, true);
```
[the remark this answers is not in the document]
Comment (bmarty) [reply matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/MXEventTimeline.java]: N/A

File: matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/MXEventTimeline.java (hunk 4 above)
Comment (bmarty) [reply matrix-sdk/src/main/java/org/matrix/androidsdk/data/timeline/MXEventTimeline.java]: N/A

File: matrix-sdk/src/test/java/org/matrix/androidsdk/data/timeline/TimelineStateHolderTest.java [hunk 7]
```diff
@@ -80,7 +80,7 @@ public void processStateEvent_WhenDirectionIsBackward__ShouldUseBackState() {
         final RoomState backState = Mockito.spy(mTimelineStateHolder.getBackState());
         mTimelineStateHolder.setBackState(backState);
         mTimelineStateHolder.processStateEvent(event, EventTimeline.Direction.BACKWARDS, false);
-        Mockito.verify(backState).applyState(mIMXStore, event, false);
+        Mockito.verify(backState).applyState(event, false, mIMXStore);
```
[the remark this answers is not in the document]
Comment (bmarty) [reply matrix-sdk/src/test/java/org/matrix/androidsdk/data/timeline/TimelineStateHolderTest.java]: done

**Pull request merged**